### PR TITLE
S2U-14 Tests & Quizzes: Canceled questions and score re-calculation

### PR DESCRIPTION
**Pull request by stetsche**
This lets instructors cancel questions, which will give them a score of 0 and scores are recalculated.

Changes to `PublishedAssessmentService`:

1. Added method `preparePublishedItemCancellation` which will process cancellation of the items
2. Moved `regradePublishedAssessment` from `RepublishAssessmentListener`

`preparePublishedItemCancellation` is used by `ItemCancellationListener`
`regradePublishedAssessment` is used by `RepublishAssessmentListener` and `ItemCancellationListener`

`ItemCancellationListener` is called from Edit assessment page - `editAssessment.jsp`, when editing a published assessment (republishing). And also from the Questions ab Published Assessment -> Scores -> Questions `questionScores.jsp`

Item cancellation works like so:
1. Is called and sets pending cancellation on item (like `ITEM_TOTAL_SCORE_TO_CANCEL`)
2. preparePublishedItemCancellation will set the score to 0 and depending on the cancellation method adjust scores of other items. The sore is set to a cancelled cancellation (like `ITEM_TOTAL_SCORE_CANCELLED`)
3. Regrading will happen when the regrade and republish button is is clicked of for the Questions tab view, right away. The regrading is just based on the new scores and not affected by the cancellation of the item.

Details on how the two cancellation methods work are detailed int the Jira:
https://sakaiproject.atlassian.net/browse/S2U-14

Conversion Scripts: https://github.com/sakaiproject/sakai-reference/pull/191

**Missing:** ES, CA and EU translations

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+        String itemId = ContextUtil.lookupParam("itemId");
+        String cancellation = ContextUtil.lookupParam("cancellation");
+        String outcome = ContextUtil.lookupParam("outcome");
+        boolean regrade = Boolean.parseBoolean(ContextUtil.lookupParam("regrade"));
+
+        AssessmentBean assessmentBean = (AssessmentBean) ContextUtil.lookupBean("assessmentBean");
+        QuestionScoresBean questionScoresBean = (QuestionScoresBean) ContextUtil.lookupBean("questionScores");
+
+        log.debug("ItemCancel LISTENER.");
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java:58]: Maybe better only one log.debug with all parameters... What do you think?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            //Add error message to context
+            String errorMessage = ContextUtil.getLocalizedString("org.sakaiproject.tool.assessment.bundle.CommonMessages",
+                    "cancel_question_error_cancelling");
+            FacesContext.getCurrentInstance().addMessage(null, new FacesMessage(errorMessage));
+
+            throw new AbortProcessingException(String.format("AssessmentId [%s], ItemId [%s] or cancellation [%s] is not parsable",
+                    publishedAssessmentId, itemId, cancellation));
+        }
+
+        // int cancellation = Integer.parseInt(cancellationString);
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java:83]: Remove

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        // Set cancellation and save item
+        publishedItemFacade.getData().setCancellation(Integer.parseInt(cancellation));
+        publishedItemService.saveItem(publishedItemFacade);
+
+        // Process cancellation
+        PublishedAssessmentIfc updatedPublishedAssessment = publishedAssessmentService.preparePublishedItemCancellation(
+                publishedAssessmentService.getPublishedAssessment(publishedAssessmentId));
+
+        // Save assessment
+        // publishedAssessmentService.(new AssessmentFacade(updatedPublishedAssessment));
```
Comment (jesusmmp) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java:96]: Remove

File: samigo/samigo-app/src/webapp/js/authoringQuestionCancellation.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+(() => {
+
+    // For B3 we still need JQuery, isolating it here
+    function openModal(modalElement) {
+        // B3 version
+        $(modalElement).modal();
+
+        // B5 version
+        // new bootstrap.Modal(modalElement).show();
```
Comment (jesusmmp) [inline samigo/samigo-app/src/webapp/js/authoringQuestionCancellation.js:14]: We can't forget fix it when merging on 24.x. :)

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml [hunk 5]
```diff
@@ -37,6 +37,7 @@
     <property name="hash" type="string" column="HASH" not-null="false" />
     <property name="itemHash" type="string" column="ITEMHASH" not-null="false" />
     <property name="isExtraCredit" type="boolean" column="ISEXTRACREDIT" not-null="false" />
+    <property name="cancellation" type="integer"><column name="CANCELLATION" default="0" not-null="true"/></property>
```
Comment (jesusmmp) [inline samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml:40]: Should be <column="CANCELLATION" instead of <column name="CANCELLATION"?

**Comment by jesusmmp** [comment]
Awesome job!! @stetsche

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java (hunk 1 above)
Comment (stetsche) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java:58]: Yeah, makes sense!

File: samigo/samigo-app/src/webapp/js/authoringQuestionCancellation.js (hunk 4 above)
Comment (stetsche) [reply samigo/samigo-app/src/webapp/js/authoringQuestionCancellation.js:14]: Yes, right!

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml (hunk 5 above)
Comment (stetsche) [reply samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml:40]: I specified it this way, to be able to set the default. I think I can't pass this on `<property>`. I will check again.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_es.properties [hunk 6]
```diff
@@ -36,3 +36,21 @@ action_save_pair=Guardar par
 delete=Borrar
 delete_attempt=Eliminar intento
 confirm_delete_attempt=\u00bfEst\u00e1 seguro que desea eliminar este env\u00edo de examen?
+
+# S2U-14
+cancel_question=Cancel question
```
[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_es.properties:41]: I think we should do translate it before we merge it. I will add some comments.

[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_es.properties:41]: Hey @mpellicer and @bgarciaentornos, could you help me with the ES translations please?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_ca.properties [hunk 7]
```diff
@@ -36,3 +36,21 @@ action_save_pair=Desa la parella
 delete=Elimina
 delete_attempt=Elimina l\u2019intent
 confirm_delete_attempt=Esteu segur de voler eliminar aquest intent?
+
+# S2U-14
```
Comment (stetsche) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_ca.properties:40]: @mpellicer and @bgarciaentornos, could you help me with the CA translations as well please?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties [hunk 8]
```diff
@@ -36,3 +36,21 @@ action_save_pair=Gorde parea
 delete=Ezabatu
 delete_attempt=Ezabatu saialdia
 confirm_delete_attempt=Ziur zaude saialdi hau ezabatu nahi duzula?
+
+# S2U-14
+cancel_question=Cancel question
```
Comment (stetsche) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties:41]: Hi @PabloSanRoman, could you help me with the EU translations please?

**Comment by stetsche** [comment]
Thanks for taking a look, @jesusmmp! I will update the PR based on your comments.

File: samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp [hunk 9]
```diff
@@ -529,7 +541,64 @@ $(window).load( function() {
...
+  <%--Question Cancellation Modal--%>
+  <div id="cancelQuestionModal" class="modal fade question-cancel-modal" tabindex="-1">
+    <h:panelGroup styleClass="modal-dialog" layout="block">
+      <h:panelGroup styleClass="modal-content" layout="block">
+        <h:panelGroup styleClass="modal-header" layout="block">
+          <button type="button" class="close" data-dismiss="modal"
+              aria-label='<h:outputText value="#{authorMessages.button_close}" />'>
+            <span class="fa fa-times" aria-hidden="true"></span>
+          </button>
+          <h2 class="modal-title">
```
[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp:554]: The heading is not empty.

File: samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp [hunk 10]
```diff
@@ -529,7 +541,64 @@ $(window).load( function() {
...
+            <f:param name="cancellation" value="10" />
+          </h:commandButton>
+          <h:commandButton styleClass="button" immediate="true" id="cancelItemDistribute" action="editAssessment"
+              value="#{commonMessages.cancel_question_distribute_points}">
+            <f:actionListener type="org.sakaiproject.tool.assessment.ui.listener.author.ItemCancellationListener" />
+            <f:param name="outcome" value="editAssessment" />
+            <f:param name="itemId" value="ITEM_ID" />
+            <f:param name="cancellation" value="20" />
+          </h:commandButton>
+          <button type="button" class="button" data-dismiss="modal">
```
[the remark this answers is not in the document]
Comment (stetsche) [reply samigo/samigo-app/src/webapp/jsf/author/editAssessment.jsp:594]: This button has an inner text.

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java [hunk 11]
```diff
@@ -36,6 +36,12 @@ public interface ItemDataIfc extends Comparable<ItemDataIfc>, java.io.Serializab
   public static final Integer ACTIVE_STATUS = 1;
   public static final Integer INACTIVE_STATUS = 0;
 
+  public static final int ITEM_NOT_CANCELED = 0;
```
Comment (mpellicer) [inline samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java:39]: Shall we use Integer instead of a basic int?

File: samigo/samigo-api/src/java/org/sakaiproject/samigo/util/SamigoConstants.java [hunk 12]
```diff
@@ -146,12 +146,22 @@ public final class SamigoConstants {
     public static final     String      EVAL_BUNDLE                                         = "org.sakaiproject.tool.assessment.bundle.EvaluationMessages";
     public static final     String      AUTHOR_BUNDLE                                       = "org.sakaiproject.tool.assessment.bundle.AuthorMessages";
 
+    /*
+     * Author Outcomes
+     */
+    public static final     String      OUTCOME_AUTHOR_EDIT_ASSESSMENT                      = "editAssessment";
```
Comment (mpellicer) [inline samigo/samigo-api/src/java/org/sakaiproject/samigo/util/SamigoConstants.java:152]: Lovely!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_ca.properties (hunk 7 above)
Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_ca.properties:40]: Sure, I'll help in a bit.

File: samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml (hunk 5 above)
Comment (stetsche) [reply samigo/samigo-hibernate/src/java/org/sakaiproject/tool/assessment/data/dao/assessment/PublishedItemData.hbm.xml:40]: Though I can't set the default on `property`, it seems there is no problem removing the default value, I am updating to the format you suggested.

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java (hunk 11 above)
Comment (stetsche) [reply samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java:39]: The switch statements don't like `Integer`'s :thinking:

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_es.properties (hunk 6 above)
[the remark this answers is not in the document]
Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_es.properties:41]: Sure thing

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/ItemContentsBean.java [hunk 13]
```diff
@@ -1717,5 +1718,13 @@ public Long getEffectiveItemId() {
 			return itemData.getItemId();
 		}
 	}
+
+	public boolean isCancelled() {
+		return this.itemData.getCancellation().intValue() != ItemDataIfc.ITEM_NOT_CANCELED;
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/ItemContentsBean.java:1723]: If using an integer, we can something like:

! ItemDataIfc.ITEM_NOT_CANCELED.equals(this.itemData.getCancellation())

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/SectionContentsBean.java [hunk 14]
```diff
@@ -677,27 +679,50 @@ public boolean getHasAttachment(){
   }
 
   public boolean getNoQuestions() {
-	  return noQuestions;
+    return noQuestions;
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/bean/delivery/SectionContentsBean.java:682]: Revert all of the format changes please, this makes compatibility harder.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+        log.debug("itemId: {}", itemId);
+        log.debug("outcome {}", outcome);
+        log.debug("regrade: {}", regrade);
+
+        String publishedAssessmentId = assessmentBean.getAssessmentId() != null
+                ? assessmentBean.getAssessmentId()
+                : questionScoresBean.getPublishedId();
+
+        // Abort if publishedAssessmentId or passed params are invalid
+        if (!NumberUtils.isParsable(publishedAssessmentId)
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemCancellationListener.java:65]: nice!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java [hunk 16]
```diff
@@ -15,11 +15,8 @@
  */
 package org.sakaiproject.tool.assessment.ui.listener.author;
 
-import java.util.Date;
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java:18]: Revert this changes please, can make compatibility harder. I know they are unused and improves the overall quality, but will make cherry-picks harder.

Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java:18]: Ahhh, it's moved logic, its fine thenm

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties (hunk 8 above)
Comment (PabloSanRoman) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties:41]: Hello @mpellicer !!
+ # S2U-14
+ cancel_question=Ezabatu galdera

**Review by mpellicer — APPROVED** [review]
Seems very nice to me, I only have concerns with the conversion scripts. 

Tremendous work @stetsche , you're a monster! It is a great feature and very well written.

File: samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java (hunk 11 above)
Comment (mpellicer) [reply samigo/samigo-api/src/java/org/sakaiproject/tool/assessment/data/ifc/assessment/ItemDataIfc.java:39]: Yeah but we can use an enum, no problem, It's fine, thanks Markus!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties (hunk 8 above)
Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties]: Thanks Pablo! Are you going to contribute translations for the rest of the strings?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java [hunk 17]
```diff
@@ -123,51 +122,12 @@ public void processAction(ActionEvent ae) throws AbortProcessingException {
 		// Update Delivery Bean
 		DeliveryBean delivery = (DeliveryBean) ContextUtil.lookupBean("delivery");
 		delivery.setPublishedAssessment(assessment);
-
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/RepublishAssessmentListener.java:126]: Revert this format.

File: samigo/samigo-services/src/test/org/sakaiproject/tool/assessment/services/ItemCancellationTest.java [hunk 18]
```diff
@@ -0,0 +1,279 @@
+/**
```
Comment (mpellicer) [inline samigo/samigo-services/src/test/org/sakaiproject/tool/assessment/services/ItemCancellationTest.java:1]: SUPREME!

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties (hunk 8 above)
Comment (PabloSanRoman) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties:41]: Hi @mpellicer , I'm sorry, the strings in the comment are not displayed, they can be seen opening the link of _eu.properties. Anyway, here you are: 

cancel_question=Ezabatu galdera
cancel_question_reduce_total=Murriztu puntuazio osoa
cancel_question_distribute_points=Modu berean banatu puntuak
cancel_question_cancel=Ez ezabatu galdera
cancel_question_info_cancelled_question=Galdera hau bertan behera geratu da eta ez du eraginik izango puntuazio osoan.
cancel_question_info_cancellation=Galdera hori bertan behera uzten baduzu kalifikazioan ez da kontuan hartuko. \
Ezabatzearen eragina bi modutan kudeatu daiteke:
cancel_question_info_distribute_points=Ezabatutako galderaren puntuak berdin banatuko dira gainerako galderen artean.\
Puntuazio osoan eraginik ez dago.
cancel_question_info_reduce_total=Ezabatutako galderaren puntuak puntuazio osotik kenduko dira.\
Gainerako galderetan eraginik ez dago. 
cancel_question_info_regrade=Galdera ezabatutakoan egin diren bidalketak berriro kalifikatuko dira.
cancel_question_info_no_undo=Ekintza hau ezin da berriz desegin!
cancel_question_info_skip_question=Galdera hau saihestu dezakezu.
cancel_question_error_cancelling=Errore bat gertatu da galdera ezabatzerakoan. Ezer ez da aldatu.
cancel_question_info_emi=Kalifikazio doiturik ez dute jasoko "Lotzeko aurreratua" motako elementuek.

Comment (mpellicer) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/bundle/CommonMessages_eu.properties:41]: No problem, thank you!
